src/geometry/manifold/derivation_bundle.lean

```diff
@@ -0,0 +1,163 @@
+/-
+Copyright © 2020 Nicolò Cavalleri. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Nicolò Cavalleri
+-/
+
+import geometry.manifold.algebra.smooth_functions
+import ring_theory.derivation
+
+/-!
+
+# Derivation bundle
+
+In this file we define the derivations at a point of a manifold on the algebra of smooth fuctions.
+Moreover we define the total bundle of derivation (although at the moment it has not been given a
+topology). Finally we define the differential of a function in terms of derivations.
+
+The content of this file is not meant to be regarded as an alternative definition to the current
+tangent bundle but rather as a purely algebraic theory that provides a purely algebraic definition
+of the Lie algebra for a Lie group.
+
+-/
+
+variables {𝕜 : Type*} [nondiscrete_normed_field 𝕜]
+{E : Type*} [normed_group E] [normed_space 𝕜 E]
+{H : Type*} [topological_space H] (I : model_with_corners 𝕜 E H)
+(M : Type*) [topological_space M] [charted_space H M]
+
+open_locale manifold
+
+namespace instances
+def smooth_functions_algebra : algebra 𝕜 C^∞⟮I, M; 𝕜⟯ := infer_instance
+attribute [instance, priority 10000] smooth_functions_algebra
+def tower : is_scalar_tower 𝕜 C^∞⟮I, M; 𝕜⟯ C^∞⟮I, M; 𝕜⟯ := infer_instance
+attribute [instance, priority 10000] tower
+def sizeof : has_sizeof (derivation 𝕜 C^∞⟮I, M; 𝕜⟯ C^⊤⟮I, M; 𝕜⟯) := infer_instance
+attribute [instance, priority 100000] sizeof
+end instances
```

Comment:
Problematic type class inference... See the discussion at https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/Tracing.20time.20of.20type.20class.20inference with @kbuzzard and @semorrison !

Reply:
I don't think you can change priorities globally here, it might mess up other things in the library. Maybe `local attribute...` is better?

Reply:
Yes that was actually what I wanted to do after our discussion on Zulip. However, there is a problem with that because if I repeat local attributes in both files, it does not work in the second file but only in the first... Would you by chance have time to give it a try? Or could you tag some expert that could understand what is going on? One of the main problems with this type class inference is that it looks for instances of `subsingleton` every time `simp` is called with lemmas having to do with derivations, until it fails or succeeds to find such an instance... Why?

Reply:
@gebner, any chance you could look at this?

Reply:
I'm not sure I can say anything helpful.
 - Adding the short-circuiting instances `smooth_functions_algebra` and `tower` doesn't seem to be so bad.
 - I'd always use `by apply_instance` instead of `infer_instance`.  `infer_instance` requires an unnecessary δ-reduction step.
 - The `has_sizeof` class is a vestigial part of lean used in the nested inductives module and the very limited default tactic for the well-founded recursion compiler.  Unfortunately there is no way to disable the automatic creation of these instances.  But at least you can speed it up a bit by adding a high-priority instance that says everything has size 0.